python/traverse.py

```diff
@@ -0,0 +1,164 @@
+    # We can just discard that.
+    if isinstance(elements[0], str):
+        del elements[0]
+
+    for element in elements:
+        for e in element.iter():
+            fn(e)
+
+
+def format_attrib_value(v: str) -> str:
+    # https://html.spec.whatwg.org/multipage/parsing.html#escapingString
+    return v.replace("&", "&amp;").replace('"', "&quot;").replace("\xa0", "&nbsp;")
+
+
+def get_source_definition(element: lxml.html.HtmlElement) -> str:
+    attributes = (
+        f'''{k}="{format_attrib_value(v)}"''' for k, v in element.attrib.items()
+    )
+    return f"<{' '.join((element.tag, *attributes))}>"
+
+
+def traverse_and_replace(
+    html: str, replace: Callable[[lxml.html.HtmlElement], ElementReplacement]
+) -> str:
+    # Initialize result and work data structures
+    result: Deque[str] = deque()
+
+    initial_list = lxml.html.fragments_fromstring(html)
+    count_stack: Deque[int] = deque([len(initial_list)])
+    work_stack: Deque[Union[str, lxml.html.HtmlElement]] = deque(reversed(initial_list))
+    tail_stack: Deque[Tuple[str, Optional[str]]] = deque()
+
+    while work_stack:
+        element = work_stack.pop()
+
+        # For just a string, append to final result
+        if isinstance(element, str):
+            result.append(element)
+
+        else:
+            new_elements = replace(element)
+
+            # Turn new_elements into a list containing we can process
+            if new_elements is None:
+                new_elements = []
+            elif isinstance(new_elements, str):
+                fragments = lxml.html.fragments_fromstring(new_elements)
+                new_elements = fragments if fragments is not None else []
+
+            if isinstance(new_elements, list):
+                # Modify count stack for new elements and decrement for element that was replaced
+                count_stack[-1] += len(new_elements) - 1
+
+                # Add element tail before processing replaced element
+                if element.tail is not None:
+                    count_stack[-1] += 1
+                    work_stack.append(element.tail)
+
+                # Extend and go to the next iteration
+                if new_elements:
+                    work_stack.extend(reversed(new_elements))
+
+                continue
+
+            if isinstance(new_elements, lxml.html.HtmlComment):
+                result.append(lxml.html.tostring(new_elements, encoding="unicode"))
+                continue
+
+            # Add opening tag and text
+            result.append(get_source_definition(new_elements))
+
+            if new_elements.text is not None:
+                result.append(new_elements.text)
+
+            # Add all children to the work stack
+            children = list(new_elements)
+
+            if children:
+                work_stack.extend(reversed(children))
+
+            count_stack.append(len(children))
+            tail_stack.append((new_elements.tag, element.tail))
```

Comment:
```suggestion
            if isinstance(new_elements, lxml.html.HtmlComment):
                result.append(lxml.html.tostring(new_elements, encoding="unicode"))
            else:
                # Add opening tag and text
                result.append(get_source_definition(new_elements))
    
                if new_elements.text is not None:
                    result.append(new_elements.text)
    
                # Add all children to the work stack
                children = list(new_elements)
    
                if children:
                    work_stack.extend(reversed(children))
    
                count_stack.append(len(children))
                tail_stack.append((new_elements.tag, element.tail))
```
I think this is all that's needed to make it play nice with comments (tried locally). There's another issue going on with how the list of elements is returned by the `fragments_fromstring` function, but that should be handled with a different fix.

Reply:
Nice, I adopted this change in 77e207b.